linkerd/service-profiles/src/split.rs

```diff
@@ -58,31 +61,67 @@ impl<N: Clone, S, Req> Clone for NewSplit<N, S, Req> {
     }
 }
 
-impl<T, N: Clone, S, Req> NewService<T> for NewSplit<N, S, Req>
+impl<T, N, S, Req> NewService<T> for NewSplit<N, S, Req>
 where
-    T: AsRef<Receiver>,
+    T: Clone,
+    for<'t> &'t T: Into<Addr> + Into<Option<Receiver>>,
+    N: NewService<(Option<Addr>, T), Service = S> + Clone,
     S: tower::Service<Req>,
+    S::Error: Into<Error>,
 {
     type Service = Split<T, N, S, Req>;
 
     fn new_service(&mut self, target: T) -> Self::Service {
-        let rx = target.as_ref().clone();
-        Split {
-            rx,
-            target,
-            new_service: self.inner.clone(),
-            rng: self.rng.clone(),
-            inner: None,
-            services: ReadyCache::default(),
-        }
+        let inner = match Into::<Option<Receiver>>::into(&target) {
+            None => {
+                trace!("Building default service");
+                Inner::Default(self.inner.new_service((None, target)))
+            }
+            Some(rx) => {
+                let mut targets = rx.borrow().targets.clone();
+                if targets.len() == 0 {
+                    targets.push(Target {
+                        addr: (&target).into(),
+                        weight: 1,
+                    })
+                }
+                trace!(?targets, "Building split service");
+
+                let mut addrs = IndexSet::with_capacity(targets.len());
+                let mut weights = Vec::with_capacity(targets.len());
+                let mut services = ReadyCache::default();
+                let mut new_service = self.inner.clone();
+                for Target { weight, addr } in targets.into_iter() {
+                    services.push(
+                        addr.clone(),
+                        new_service.new_service((Some(addr.clone()), target.clone())),
+                    );
+                    addrs.insert(addr);
+                    weights.push(weight);
+                }
```

Comment:
i guess the only real change in this file besides adding the default case is that we now call `new_service` eagerly in `new_service` on splits, instead of doing it lazily in the first `poll_ready`? 

this _seems_ better regardless; less moving parts and `new_service` is expected to be called eagerly. i can't imagine anything relying on it happening in `poll_ready`?

Reply:
Yeah, that's right